aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -161,6 +161,17 @@ Tensor& div_sparse_(Tensor& self, const Tensor& value) {
   return div_out_sparse_zerodim(self, self, value);
 }
 
+// inplace coalesce
+const SparseTensor& coalesce_(const SparseTensor& tensor) {
+  SparseTensor coalesced = tensor.coalesce();
```

Comment:
In-place sparse division doesn't actually save any memory because we are coalescing out-of-place here (the point of in-place operations is to save on memory usage). This would be good to document in the sparse docs

Reply:
right, the purpose here is just to be able to have the change affect the reference. My intent wasn't to expose this to users.

Reply:
My suggestion was not to document "coalesce_" in the docs; it is to document to users that "in-place sparse division doesn't save any memory in its current implementation so there is no reason to use it over regular sparse division"

Reply:
ah OK, though we only use extra memory in the case where the input tensor is an uncoalesced integral tensor. That seems a bit narrow to be explaining the implementation detail.

Reply:
[is_coalesced](https://pytorch.org/docs/stable/sparse.html?highlight=coalesce#torch.sparse.FloatTensor.is_coalesced) is a public property and it's very easy to get an uncoalesced tensor when working with sparse tensors (i.e., sparse addition). I wouldn't call this an implementation detail; it is part of the API that an in-place operation saves memory. 

It is important for us to note places where the behavior might differ from what the user is expecting, either by documenting it correctly or filing an issue to be fixed.

Reply:
Is there somewhere in the docs that already goes into this level of detail about the memory tradeoffs of in-place ops and impact of uncoalesced tensors? It seems like it would be odd to document to this level of detail just for this particular change without the rest of the context you refer to. If adding the rest of the context, I'd prefer to use a separate PR. 

The only explanation I could find was the note here:
https://pytorch.org/docs/stable/sparse.html
which is pretty general, "For the most part, you shouldn’t have to care whether or not a sparse tensor is coalesced or not..."

Reply:
> Is there somewhere in the docs that already goes into this level of detail about the memory tradeoffs of in-place ops and impact of uncoalesced tensors? 

It looks like we don't document the semantics of "in-place" operations. However, users use those as a strategy to save memory.

See
[this](https://towardsdatascience.com/in-place-operations-in-pytorch-f91d493e970e), and [this](https://discuss.pytorch.org/t/what-is-in-place-operation/16244) (a few things I found from a quick search)

> The only explanation I could find was the note here:
https://pytorch.org/docs/stable/sparse.html
which is pretty general, "For the most part, you shouldn’t have to care whether or not a sparse tensor is coalesced or not.."

I don't know if `div_out` saves memory. If it does, and if we are adding a new case where it doesn't save memory, then saying that a "user shouldn't care whether or not a sparse tensor is coalesced" means that we should not be inconsistent over whether or not `div_out` saves memory or not.

Reply:
> I don't know if div_out saves memory.

It saves memory in the particular case of in-place division (e.g. sparse_tensor /= 2.0) on coalesced or floating tensors, but I don't see the general characteristics being a documented anywhere yet (i.e. this change doesn't make any existing documentation wrong). I'm interpreting "user shouldn't care" as meaning it doesn't change the semantics of their code -- obviously they do sometimes already care since it does impact performance (regardless of this current PR).

It could be helpful to document the memory characteristics of sparse and uncoalesced ops in more detail than we have, but that seems outside the scope of this PR.

I'm not entirely convinced that we should go into too much detail when/if we do, since to do so would be documenting performance characteristics that could frequently change or regress. E.g. I see this comment in `SparseCUDATensorMath.cu`:
```
  // FIXME: add some heuristic about when to call coalesce() here, so that
  // tensors don't totally blow up in size by concatenation; e.g.
  //   if (r->nnz / r->minUnique > COMPACTION_THRESHOLD) {
  //     THCSTensor_(contiguous)(r);
  //     r->minUnique = r->nnz;
  //   }
```
If we try to fully document the impact of using uncoalesced sparse tensors in arithmetic ops, that kind of heuristic seems a bit much for user documentation. I understand that users care about performance, but this seems like something where approaches should be explained generally at a high level, or maybe point to blog posts, not necessarily detailing all cases where use of an uncoalesced tensor inside an arithmetic op might lead to additional memory overhead.

Reply:
Yes, you've convinced me that differentiating between coalesced and uncoalesced tensors in the documentation is a bad idea.

However, I still think that we need a note somewhere that in-place sparse division may use more memory than expected -- we could just write "in some cases, sparse in-place division may perform a copy_ of the tensor and not save memory". Just because we have not documented something explicitly doesn't mean that our users don't expect the behavior. As a user of pytorch, I expect in-place operations to not use additional memory.